internal/runtime/hcsv2/nvidia_utils.go

```diff
@@ -0,0 +1,77 @@
+// +build linux
+
+package hcsv2
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/Microsoft/opengcs/internal/storage/pci"
+	oci "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+const lcowNvidiaMountPath = "/run/nvidia"
+
+// addNvidiaDevicePreHook builds the arguments for nvidia-container-cli and creates the prestart hook
+func addNvidiaDevicePreHook(ctx context.Context, spec *oci.Spec) error {
+	nvidiaToolBinary := "nvidiaPrestartHook"
+	nvidiaToolPath, err := exec.LookPath(nvidiaToolBinary)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find %s for container GPU support", nvidiaToolBinary)
+	}
+
+	var args []string
+	args = append(args, nvidiaToolPath, "--debug=/nvidia-container.log", "--load-kmods", "--no-pivot", "configure", "--ldconfig=/sbin/ldconfig")
+	if capabilities, ok := spec.Annotations["io.microsoft.container.gpu.capabilities"]; ok {
+		caps := strings.Split(capabilities, ",")
+		for _, c := range caps {
+			args = append(args, fmt.Sprintf("--%s", c))
+		}
+	}
+
+	for _, d := range spec.Windows.Devices {
+		switch d.IDType {
+		case "gpu":
+			busLocation, err := pci.FindDeviceBusLocationFromVMBusGUID(ctx, d.ID)
+			if err != nil {
+				return errors.Wrapf(err, "failed to find nvidia gpu bus location")
+			}
+			args = append(args, fmt.Sprintf("--device=%s", busLocation))
+		}
+	}
+
+	rootfsDir := spec.Root.Path
+	args = append(args, "--no-cgroups", "--pid=%v", fmt.Sprintf("%s", rootfsDir))
+
+	if spec.Hooks == nil {
+		spec.Hooks = &oci.Hooks{}
+	}
+
+	nvidiaHook := oci.Hook{
+		Path: nvidiaToolPath,
+		Args: args,
+		Env:  updateEnvWithNvidiaPath(),
+	}
+
+	spec.Hooks.Prestart = append(spec.Hooks.Prestart, nvidiaHook)
+	return nil
+}
+
+// updateEnvWithNvidiaPath creates an env with nvidia bits in PATH and insecure mode set
+func updateEnvWithNvidiaPath() []string {
+	pathPrefix := "PATH="
+	nvidiaBin := fmt.Sprintf("%s/bin", lcowNvidiaMountPath)
+	env := os.Environ()
+	for i, v := range env {
+		if strings.HasPrefix(v, pathPrefix) {
+			newPath := fmt.Sprintf("%s:%s", v, nvidiaBin)
+			env[i] = newPath
+		}
+	}
+	env = append(env, "NVC_INSECURE_MODE=1")

```

Comment:
What is this?

Reply:
nvidia-container-cli typically uses seccomp to limit syscalls. In order to prevent using seccomp, the tool expects this environment variable to [be set](https://github.com/NVIDIA/libnvidia-container/blob/6ed0f129eb63142ce60a57e7657638b552327a85/src/nvc_ldcache.c#L41) as well as the package to have been [compiled without seccomp support](https://github.com/NVIDIA/libnvidia-container/blob/6ed0f129eb63142ce60a57e7657638b552327a85/Makefile#L12)

Reply:
Cool. Let's add a comment here saying that, and rename the function this is in so it's clear it's doing more than just setting PATH.

Reply:
Done!